lightning/src/util/events.rs

```diff
@@ -1422,19 +1446,12 @@ impl MaybeReadable for Event {
 					let mut failed_next_destination_opt = None;
 					read_tlv_fields!(reader, {
 						(0, prev_channel_id, required),
-						(2, failed_next_destination_opt, ignorable),
+						(2, failed_next_destination_opt, upgradable_required),
 					});
-					if let Some(failed_next_destination) = failed_next_destination_opt {
-						Ok(Some(Event::HTLCHandlingFailed {
-							prev_channel_id,
-							failed_next_destination,
-						}))
-					} else {
-						// If we fail to read a `failed_next_destination` assume it's because
-						// `MaybeReadable::read` returned `Ok(None)`, though it's also possible we
-						// were simply missing the field.
-						Ok(None)
-					}
+					Ok(Some(Event::HTLCHandlingFailed {
+						prev_channel_id,
+						failed_next_destination: _init_tlv_based_struct_field!(failed_next_destination_opt, upgradable_required),
```

Comment:
Shouldn't the early return in `_init_tlv_based_struct_field` already be hit in `_decode_tlv` via `read_tlv_fields` above?

Reply:
Removed the early return from `_init_tlv_based_struct_field`

Reply:
Why not `unwrap` instead of needing to use `_init_tlv_based_struct_field`? Seems strange that this is a requirement for using `upgradable_required`.

Reply:
It's not a requirement, we could `unwrap`. One reason to use it is that In the past, not using `_init_tlv_based_struct_field` hid a bug in our ser macros where `ignorable` was treated differently when used in lower-level macros (eg `read_tlv_stream`) vs higher-level macros (eg `impl_writeable_*`) (see https://github.com/lightningdevkit/rust-lightning/pull/2014#discussion_r1110319709). So this ensures they're consistent and also means you don't need to check an `unwrap` for safety.

Reply:
Well, using `_init_tlv_based_struct_field` in lower level macros would've caught the inconsistency, so it's nice to know it works now. And yeah there's always an unwrap but this way you can just check that both ser macros specify `upgradable_required` instead of checking the unwrap itself, but, not a huge deal there.